_posts/2023-11-17-transactional-outbox.md

```diff
@@ -0,0 +1,295 @@
+and then run the Spring Boot application.
+2. **Creating Users:** To test user creation and the subsequent event publishing, you can use the `/users` endpoint. 
+This will trigger the flow from the user creation in the `UserService` to the publishing and handling of the `UserCreatedEvent`.
+3. **Observing Event Handling:** As you create users through the endpoint, keep an eye on the logs of the RabbitMqConsumer. 
+You should see the `UserCreatedEvent` payloads being logged, indicating that the events are successfully being consumed from RabbitMQ.
+
+For convenience, the repository includes a `sample-requests.http` file,
+which provides easy access to all the endpoints if you are using IntelliJ IDEA.
+This file simplifies the process of making requests to the application for testing purposes.
+
+## Explore the Code
+
+The `UserController` controller provides an endpoint to create users.
+```kotlin
+@PostMapping("/users")
+fun createUser(@RequestBody request: UserCreateRequestDto): User {
+    return userService.createUser(request.name)
+}
+```
+
+When a user is created, a `UserCreatedEvent` is published. 
+This is done using the OutboxPublisher to persist the event in the outbox first, and then forward it to RabbitMQ.
+```kotlin
+class UserService(
+    private val userRepository: UserRepository, 
+    private val eventPublisher: EventPublisher) {
+    
+    fun createUser(name: String): User {
+        return userRepository.save(User(UUID.randomUUID(), name)).also {
+            eventPublisher.publish(UserCreatedEvent(it.id, it.name))
+        }
+    }
+}
+
+class OutboxPublisher(private val outbox: TransactionOutbox): EventPublisher {
+    override fun publish(event: DomainEvent) {
+        outbox.schedule(RabbitMqSender::class.java).send(event)
+    }
+}
+```
+
+The `schedule` method leads to the persisting of the task, which involves sending the event to RabbitMQ. 
+This task is stored in the outbox. 
+Then, it schedules the execution of this task for after the current transaction has been committed. 
+This ensures that the event is only sent after the transaction is successfully completed.
+
+In the context of handling potential failures during event sending after a transaction, the `TransactionalOutboxScheduler` class provides a robust fallback mechanism. 
+This class periodically executes a run method that continuously calls `transactionOutbox.flush()`.
+The `flush` method in `TransactionOutbox` is responsible for finding and retrying events that haven't been successfully executed after a certain amount of time. 
+It specifically targets events that were initially set up to be executed but haven't been completed as expected and have not exceeded a predefined number of retry attempts.
+
+```kotlin
+class TransactionalOutboxScheduler(private val transactionOutbox: TransactionOutbox) {
+    @Scheduled(fixedDelay = 2000)
+    fun run() {
+        while (transactionOutbox.flush()) {
```

Comment:
code review: this code smells.
 many questions arise: is it async and you have to wait? or does it retry in case of failure? or does it flush one event after another? what is this loop needed for? please make it more obvious, event if it's less elegant. maybe like this? (hard to tell because I don't know why the loop exists)

```
while (hasAnotherEvent()) flushNextEvent()

// or
while (eventsNotFlushedYet()) Thread.sleep(5 ms) // or await Async.Delay(5)?

// or
await flushEvents()

// or
while (true)
   try 
      attemptFlushEvents(); 
      break loop
   catch { /* try again next iteration, */ }

```

Reply:
The transactionOutbox.flush() is a blocking call that process all relevant events. It returns true as long as any event has been processed. Since it is a library implementation, the details are out of our scope. We only know whether work has been done, but not whether there are still events in the table that must be processed.

Reply:
well, if you show this code in the blog post and expect us to read it, we will want to know what this code does.

maybe 
```
do {  val someEvensGotFlushed = outbox.flush(); } 
while (someEventsGotFlushed);
```

Reply:
The purpose of the loop in the run method is to continuously attempt to flush transactions from the transactionOutbox. This is necessary because the flush method might not be able to process all pending tasks in a single invocation, especially in high-load scenarios.

Moreover, we are dealing with a complex topic here. The introduction of a variable, in this context, does not sufficiently explain the complexity or the nuances of the transaction outbox pattern. In fact, it might lead to more questions about the specifics of the events being flushed - a topic that is better understood through comprehensive documentation of the flush method and the transactional outbox pattern. But that goes beyond the scope of this article, since one has to understand the implementation details of this specific library.

Reply:
If we don't have to understand the code, why show the code in the first place?

Reply:
I added some prose to make it more clear.